azurelinuxagent/common/protocol/wire.py

```diff
@@ -843,25 +841,24 @@ def get_ext_conf(self):
                 self.ext_conf = ExtensionsConfig(xml_text)
         return self.ext_conf
 
-    def get_ext_manifest(self, ext_handler, goal_state):
-        for update_goal_state in [False, True]:
+    def get_ext_manifest(self, ext_handler, incarnation):
+        local_file = MANIFEST_FILE_NAME.format(
+                        ext_handler.name,
+                        incarnation)
+        local_file = os.path.join(conf.get_lib_dir(), local_file)
+        try:
+            xml_text = self.fetch_cache(local_file)
+        except ProtocolError:
+            # Either I/O error while reading the cache or the cache
+            # does not exist. Either way, grab a fresh copy.
             try:
-                if update_goal_state:
-                    self.update_goal_state(forced=True)
-                    goal_state = self.get_goal_state()
-
-                local_file = MANIFEST_FILE_NAME.format(
-                                ext_handler.name,
-                                goal_state.incarnation)
-                local_file = os.path.join(conf.get_lib_dir(), local_file)
                 xml_text = self.fetch_manifest(ext_handler.versionUris)
-                self.save_cache(local_file, xml_text)
-                return ExtensionManifest(xml_text)
-
             except ResourceGoneError:
-                continue
+                raise RestartError("Failed to retrieve extension manifest")
+
+            self.save_cache(local_file, xml_text)
```

Comment:
I think this idea is fine, but I dislike the change in caching pattern for the extension config specifically.

Reply:
Could you please elaborate on what you mean by this? What is the (old/new) caching pattern for the extension config?

Reply:
Old pattern: use cached manifest, if download fails, update goal state and retry
New pattern: use cached manifest, if download fails, throw, pass up the stack, then retry.

Can you clarify the benefit of this approach?

Reply:
Let's say there are 5 extensions to process - A through E.
Let's say that the manifest fails to download the manifest for extension C.

Old pattern: Process extensions A & B with Goal State N. Process extensions C through E with Goal State N+1. If Goal State N+2 or later shows up when the loop comes back around to processing extensions, the extensions aren't processed according to Goal State N+1.

New pattern: Process extensions A& B with Goal State N. Exit before processing extension C because fetching the manifest failed. Process the extensions as listed in Goal State N+1.